### PR TITLE
build: upgrade to latest REVM

### PR DESCRIPTION
**Pull request by Wodann**
This PR bumps our REVM dependency to https://github.com/Wodann/revm/commit/31771cca888708b09f722b0b2721ede221d383dd, which is the latest commit of REVM with some minor modifications:
- Revert of a yanked dependency
- Splitting of the `trait EvmWiring` into an additional `trait ChainSpec` such that we can have a compile-time constant `ChainSpec` without depending on the `Database` and `ExternalContext` associated types.
- Revert changes to REVM that require Rust v1.81.

**NOTE**
Support for Rust v1.81 is blocked by [a bug](https://github.com/rust-lang/rust/issues/130210) introduced in Rust v1.81. Unfortunately, REVM requires v1.81 without the changes made in our [fork](https://github.com/Wodann/revm/commit/f2600745f9f2c8c62af710393a5c0a8dc3461b07) and [EDR](https://github.com/NomicFoundation/edr/pull/669/commits/60754809f16f3f18f138175ae7a5fec3f5042937), preventing us from reverting to an older version of Rust.

<details>

<summary>Compiler errors in REVM using Rust v1.80 - without changes</summary>
```
error[E0658]: use of unstable library feature 'error_in_core'
  --> crates/primitives/src/eip7702/bytecode.rs:96:6
   |
96 | impl core::error::Error for Eip7702DecodeError {}
   |      ^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
   --> crates/primitives/src/bytecode.rs:238:6
    |
238 | impl core::error::Error for BytecodeDecodeError {}
    |      ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
   --> crates/primitives/src/bytecode/eof.rs:185:6
    |
185 | impl core::error::Error for EofDecodeError {}
    |      ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
  --> crates/primitives/src/evm_wiring.rs:21:35
   |
21 |     type ValidationError: Debug + core::error::Error;
   |                                   ^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
   --> crates/primitives/src/precompile.rs:143:6
    |
143 | impl core::error::Error for PrecompileErrors {}
    |      ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
   --> crates/primitives/src/precompile.rs:198:6
    |
198 | impl core::error::Error for PrecompileError {}
    |      ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
   --> crates/primitives/src/result.rs:316:6
    |
316 | impl core::error::Error for InvalidTransaction {}
    |      ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information

error[E0658]: use of unstable library feature 'error_in_core'
   --> crates/primitives/src/result.rs:392:6
    |
392 | impl core::error::Error for InvalidHeader {}
    |      ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #103765 <https://github.com/rust-lang/rust/issues/103765> for more information
```

</details>

### CODE REVIEW

**Review by agostbiro — APPROVED** [review]
Thanks, LGTM

**Review by Xanewok — APPROVED** [review]
Thanks for simplifying some of the traits/ADTs! The changes make sense and strictly improve what we have, so I'm fine with landing this if that's possible, left one thread about the naming but that's non-blocking.

For the Rust 1.81, would it be possible to submit a patch that downgrades MSRV for revm to use the `std` Error types for now? I looked at the upstream rustc issue and I'm not sure that's going to be backported soon, so it looks like we're stuck on <1.81 for a while.

File: crates/edr_optimism/Cargo.toml [hunk 1]
```diff
@@ -4,23 +4,24 @@ version = "0.3.5"
...
+alloy-serde = { version = "0.3.5", default-features = false, features = ["std"] }
 edr_eth = { path = "../edr_eth", features = ["serde", "std"] }
 edr_evm = { path = "../edr_evm" }
 edr_generic = { path = "../edr_generic" }
 edr_napi_core = { path = "../edr_napi_core" }
 edr_provider = { path = "../edr_provider" }
 edr_rpc_eth = { path = "../edr_rpc_eth" }
 log = { version = "0.4.17", default-features = false }
-revm = { git = "https://github.com/Wodann/revm", rev = "a500675", version = "12.1", default-features = false, features = ["c-kzg", "dev", "optimism", "serde", "std"] }
+revm = { git = "https://github.com/Wodann/revm", rev = "f260074", version = "14.0", default-features = false, features = ["c-kzg", "dev", "serde", "std"] }
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_optimism/Cargo.toml:17]: As mentioned in the description of the PR, some other changes were required that are also not merged in REVM yet.

[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/edr_optimism/Cargo.toml:17]: Ah, you're right - I somehow missed it in the OP. Thanks for clarifying!

File: crates/edr_evm/src/chain_spec.rs [hunk 2]
```diff
@@ -24,15 +27,16 @@ use crate::{
 /// A trait for defining a chain's associated types.
 // Bug: https://github.com/rust-lang/rust-clippy/issues/12927
 #[allow(clippy::trait_duplication_in_bounds)]
-pub trait ChainSpec:
+pub trait EvmSpec:
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/edr_evm/src/chain_spec.rs:30]: Makes sense. Yeah, I think RuntimeSpec does sound better in this case. Thanks for being open and sorry for the possible churn; again, don't feel like you're blocked on this :pray:

[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm/src/chain_spec.rs]: Addressed in https://github.com/NomicFoundation/edr/pull/669/commits/6d0ed6dba22a1c36734e545fa6fa1d4651da3124

**Pull request merged**